src/main/java/db/migration/V194__Migrate_instructor_userRoles.java

```diff
@@ -0,0 +1,156 @@
+                    // senateInstructor: 1
+                    // lecturer: 14
+                    if (roleId != 8 && roleId != 1 && roleId != 14) {
+                        continue;
+                    }
+
+                    userRolesDTO.workgroupIdsForUser.add(workgroupId);
+
+                    RolesDTO rolesDTO = userRolesDTO.rolesForWorkgroup.get(workgroupId);
+
+                    if (rolesDTO == null) {
+                        userRolesDTO.rolesForWorkgroup.put(workgroupId, new RolesDTO());
+                        rolesDTO = userRolesDTO.rolesForWorkgroup.get(workgroupId);
+                    }
+
+                    switch (roleId) {
+                        case 8: // FederationInstructor role
+                            rolesDTO.isFederation = true;
+                            break;
+                        case 1: // SenateInstructor role
+                            rolesDTO.isSenate = true;
+                            break;
+                        case 14: // Lecturer role
+                            rolesDTO.isLecturer = true;
+                            break;
+                        default:
+                            break;
+                    }
+                }
+
+                // Create instructor userRoles based on the old userRoles
+                for (Long workgroupId: userRolesDTO.workgroupIdsForUser) {
+                    // We need to make one userRole for each workgroup the user is an instructor in
+                    Long roleId = 15L; // New role 'instructor'
+
+                    // Identify the instructorType to use in the new userRole
+                    Long instructorTypeId = null;
+
+                    RolesDTO rolesDTO = userRolesDTO.rolesForWorkgroup.get(workgroupId);
+
+                    // If senate/federation and lecturer are present, ignore lecturer role as it is less precise.
+                    // Cannot be senate and federation simultaneously for a workgroup.
+                    if (rolesDTO.isSenate) {
+                        instructorTypeId = 6L;
+                    } else if (rolesDTO.isFederation) {
+                        instructorTypeId = 4L;
+                    } else if (rolesDTO.isLecturer) {
+                        instructorTypeId = 7L;
+                    }
+
+                    // Ensure UserRole does not already exist
+                    Long existingUserRoleId = null;
+
+                    PreparedStatement psExistingUserRole = connection.prepareStatement(
+                            "SELECT * FROM UserRoles WHERE UserId = ? AND WorkgroupId = ? AND RoleId = ?;"
+                    );
+                    psExistingUserRole.setLong(1, userId);
+                    psExistingUserRole.setLong(2, workgroupId);
+                    psExistingUserRole.setLong(3, roleId);
+                    ResultSet rsExistingUserRole = psExistingUserRole.executeQuery();
+
+                    while(rsExistingUserRole.next()) {
```

Comment:
Why loop over all possible results when you're only taking the last one?

Reply:
Can you fix this? If you're only taking the last one, you don't need the while loop.